### PR TITLE
fv3fit: generalize interface of emulator loss functions

### PR DESCRIPTION
**Pull request by nbren12**
Keras' .compile function is poorly documented for multiple output problems.
Also, for more complex loss functions (e.g. temperature scaled loss), this
interface is not suitable.

* refactor loss functions to have the interface `loss, info = loss_fn(data, prediction)`
* remove the unused StandardLoss object
* move loss functions to `fv3fit.emulation.losses`
* `fv3fit.emulation.train`: separate loss and optimzer into separate args

### CODE REVIEW

File: external/fv3fit/fv3fit/emulation/losses.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            use for the MSE
+        loss_variables: variable names to include in the MSE loss dict
+        metric_variables: variable names to include in the metrics dict
+        weights: custom scaling for the loss variables applied in the
+            overall keras "loss" term
+    """
+
+    optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/losses.py:43]: Remove optimizer from the dataclass

File: external/fv3fit/fv3fit/emulation/losses.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        loss = 0.0
+        for out_varname in x:
+            if out_varname in self.loss_variables + self.metric_variables:
+                loss_value = self.loss_funcs[out_varname](
+                    x[out_varname], y[out_varname]
+                )
+                weight = self.weights.get(out_varname, 1.0)
+                if out_varname in self.loss_variables:
+                    loss += loss_value * weight
+                metrics[out_varname] = loss_value
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/losses.py:81]: Nice! So analagously, everything that's in metrics *or* loss will still have an individual value tracked.

File: external/fv3fit/fv3fit/emulation/trainer.py [hunk 3]
```diff
@@ -12,10 +15,31 @@ def call(self, x):
...
 
 
+class _LayerTrainer(tf.keras.Model):
+    """A class for training multiple output layers
+    
+    This uses the more explicit `add_loss` API for clarity
+    """
+
+    def __init__(self, model: tf.keras.Model, loss: LossFunction):
+        # TODO doesn't work for "StandardLoss"...can we delete that one?
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/trainer.py:25]:
```suggestion
```

File: external/fv3fit/fv3fit/emulation/trainer.py [hunk 4]
```diff
@@ -63,25 +88,15 @@ def train(
...
     # This minimizes our contact points to keras APIs...which are often
     # poorly documented
     return wrapped_model.fit(
-        dataset.map(split_in_out),
+        dataset,
         epochs=epochs,
-        validation_data=validation_data.map(split_in_out)
-        if validation_data is not None
-        else None,
+        validation_data=validation_data if validation_data is not None else None,
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/trainer.py:99]:
```suggestion
        validation_data=validation_data,
```

**Review by frodre — COMMENTED** [review]
Looks good and +1 to making it use a more clearly defined keras API since we're on dictionaries now. 

Would be nice to see an example wandb training just to verify that we still get the output metrics tracked as expected.  Also, not sure if the hypothesis failure is something that is actually broken with the unit tests.

File: external/fv3fit/fv3fit/emulation/losses.py (hunk 1 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/losses.py:43]: I wanted to preserve backwards compatibility for now...so didn't change this interface. I'll probably separate the config from the `__call__`  implementation in a future PR.

File: external/fv3fit/fv3fit/emulation/losses.py (hunk 2 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/losses.py:81]: Yes...I believe so. I tried to leave this behavior unchanged.

Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/losses.py:81]: Here's a run with the new model: https://wandb.ai/ai2cm/microphysics-emulation/runs/30u6vfcn

Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/losses.py:81]: Actually, I needed to append `_loss` to make the names unchanged: https://wandb.ai/ai2cm/microphysics-emulation/runs/3iijf23y/logs?workspace=user-nbren12.

Comment (frodre) [reply external/fv3fit/fv3fit/emulation/losses.py]: Thanks for checking it.

File: external/fv3fit/fv3fit/emulation/losses.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        for out_varname in x:
+            if out_varname in self.loss_variables + self.metric_variables:
+                loss_value = self.loss_funcs[out_varname](
+                    x[out_varname], y[out_varname]
+                )
+                weight = self.weights.get(out_varname, 1.0)
+                if out_varname in self.loss_variables:
+                    loss += loss_value * weight
+                # append "_loss" for backwards compatibility
+                metrics[out_varname + "_loss"] = loss_value
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/losses.py:82]: It's probably okay to have everything suffixed for simplicity, but it might be confusing down the line if we have metric variables that are not included in the loss labeled as such.

**Review by frodre — APPROVED** [review]
LGTM and thanks for the clarifications.  Just fix up the test function w/ the suffixes and should be good to go.

File: external/fv3fit/fv3fit/emulation/losses.py (hunk 5 above)
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/losses.py:82]: Agreed. I think we may want to remove the "metric_variables" config and replace it with separate "metrics" functions with a similar interface.
